cmd/object-handlers.go

```diff
@@ -152,8 +176,72 @@ func (api objectAPIHandlers) GetObjectHandler(w http.ResponseWriter, r *http.Req
 		length = hrange.getLength()
 	}
 
+	getObject := objectAPI.GetObject
+	if api.CacheAPI() != nil && !hasSSECustomerHeader(r.Header) {
+		getObject = api.CacheAPI().GetObject
+	}
+
 	var writer io.Writer
-	writer = w
+
+	if objInfo.IsCompressed() {
+		var pipeErr error
+
+		// Defaulting the length and startOffset.
+		// Incase of range based queries, the entire data is read to the pipeWriter.
+		// So that range is set on the decompressed data.
+		snappyStartOffset := startOffset
+		snappyLength := length
+		length = compressedSize
+		startOffset = 0
+
+		// Validate the range.
+		if hrange != nil {
+			// For negative length we read everything.
+			if snappyLength < 0 {
+				snappyLength = actualSize - snappyStartOffset
+			}
+
+			// Reply back invalid range if the input offset and length fall out of range.
+			if snappyStartOffset > actualSize || snappyStartOffset+snappyLength > actualSize {
+				writeErrorResponse(w, ErrInvalidRange, r.URL)
+				return
+			}
+		}
+		objInfo.Size = snappyLength
+		setObjectHeaders(w, objInfo, hrange)
+		setHeadGetRespHeaders(w, r.URL.Query())
+
+		// Open a pipe for compression
+		// Where pipeWriter is actually passed to the getObject
+		pipeReader, pipeWriter := io.Pipe()
+		snappyReader := snappy.NewReader(pipeReader)
+
+		go func() {
+			wg.Add(1)
```

Comment:
Never call `Add` inside go-routine - It's not guaranteed that `wg.Wait()` will wait for go-routine b/c it may never be executed.

Reply:
Okay will change this !